frontend/docsettings.lua

```diff
@@ -176,8 +211,17 @@ function DocSettings:open(doc_path)
     end
     local history_file = new:getHistoryPath(doc_path)
 
+    if lfs.attributes(DOCSETTINGS_HASH_DIR, "mode") == "directory" then
```

Comment:
Not a fan of the extra stat here, I'd just cache this at require time, runtime changes be damned.

Reply:
This is here to only run `new:getSidecarHashDirAndFilepath(doc_path)` if the outer directory exists to avoid calculating a hash otherwise. If we put this check at require time, would things still work after a user changed their SDR preferred 'location' to hash-based but hadn't restarted the app yet?

Reply:
Sorry, don't understand.

Reply:
> would things still work after a user changed their SDR preferred 'location' to hash-based and maybe opened and closed some documents but hadn't restarted the app yet?

It would keep calculating hashes until a restart, yes, hence my "runtime changes be damned".

That's a better trade-off to me than always eating a `stat` on open when 99.99% of users will never use this mode ;).

Reply:
..Should we prompt for a restart then?

Reply:
Yes, I saw the popup function! I'll figure out how to add it, hopefully it actually works since I can't get UIManager:show() working above. I'm fine with this.

Reply:
Wait a moment.
sdr location for previously opened books is not changed automatically when changing the setting.
It is changed when we open a book, when we change the status, or when we call "Move metadata".
Assume you have old books with the "hash" location.
With above "restart" proposition, if your current setting is "doc" or "dir" these books will not be shown as "opened".

Reply:
Updated

Reply:
> than always eating a `stat` on open when 99.99% of users will never use this mode

Maybe, return the checkbox "Hash location enabled"?

Reply:
> Assume you have old books with the "hash" location.
> With above "restart" proposition, if your current setting is "doc" or "dir" we will not be able to open these books with their old settings

https://github.com/koreader/koreader/pull/10945/files#diff-e8930dfe3beade539965223b8d896677cae3de5ea9f57717fe02a02a1a579930R21 allows reading from old 'hash-based' book metadata irrespective of the current SDR location. So after a restart, if there are still said old books, then I think their metadata should still be readable (and hopefully would be automoved). Once there is no hash-based metadata remaining on restart, only then will the hash directory not be checked.

My concern for restarting was more because without a restart, if a user changes to the hash-based SDR setting for the first time (hash dir is initially empty) and then opens/closes several books without a restart, the metadata that was saved into the hash-dir before the restart would not be readable until after a restart.

Reply:
> > than always eating a `stat` on open when 99.99% of users will never use this mode
> 
> Maybe, return the checkbox "Hash location enabled"?

That would make perfect sense, yeah ;).

Reply:
> That would make perfect sense, yeah ;).

Assuming nobody cares about even attempting to look at them if it's not enabled? (i.e., switching from a "live in" hash mode to something else would essentially ignore these old hashed docsettings).

Reply:
(haven't read the code nor anything much...)
store the state of this as a module var nil/true/false
If nil, check if dir exist => true or false
When in the menu changeing the type of docsetting storage, reset this to nil.
?

Reply:
That'd work if there was only a single dir to check, and it was always the same, I think?

Reply:
> If we do not expose enabling hash-based location, a user could have just one book with hash-based sdr (got it when tested a new feature), then switched to "doc" and forgot about it. And will calculate hashes forever.

@hius07's comment was outside this thread but I'm fine to put in the checkbox if this is a concern. Do you by chance already have some source code / commit to share based on your screenshots? It would be helpful (especially the greyed out thing which I don't know how to do) to speed up development.

Reply:
This has been implemented - thanks for the code headstart @hius07.